Microsoft.Exchange.WebServices.Tests/WithDnsClient.cs

```diff
@@ -0,0 +1,121 @@
+using System.Linq;
+using System.Net;
+using AutoFixture;
+using AutoFixture.AutoMoq;
+using DnsClient;
+using DnsClient.Protocol;
+using Microsoft.Exchange.WebServices.Dns;
+using Moq;
+using NUnit.Framework;
+
+namespace Microsoft.Exchange.WebServices.Tests;
+
+[TestFixture]
+public class WithDnsClient
+{
+    private IFixture fixture;
+    private Dns.DnsClient dnsClient;
+
+    private Mock<IDnsQueryFactory> dnsQueryFactoryMock;
+    private Mock<IDnsQueryResponse> dnsQueryResponseMock;
+
+    private static IPAddress[] possibleDnsServerIpAddresses = { IPAddress.Parse("1.1.1.1"), null };
+
+    [SetUp]
+    public void SetUp()
+    {
+        fixture = new Fixture().Customize(new AutoMoqCustomization { ConfigureMembers = true });
+
+        dnsQueryFactoryMock = fixture.Freeze<Mock<IDnsQueryFactory>>();
+        dnsQueryResponseMock = fixture.Freeze<Mock<IDnsQueryResponse>>();
+        dnsClient = new Dns.DnsClient(dnsQueryFactoryMock.Object);
+
+        fixture.Register<DnsResourceRecord>(() => fixture.Create<SrvRecord>());
+        SetDnsQueryResponseError(errorMessage: null);
+    }
+
+    [TestCaseSource(nameof(possibleDnsServerIpAddresses))]
+    public void SrvQuery_Should_Pass_DnsServerAddress_To_DnsQueryFactory(IPAddress dnsServerAddress)
+    {
+        // Act
+        dnsClient.SrvQuery(domain: "any.domain", dnsServerAddress);
+
+        // Assert
+        dnsQueryFactoryMock.Verify(
+            dnsQueryFactory => dnsQueryFactory.GetDnsQueryClient(dnsServerAddress),
+            Times.Once);
+        dnsQueryFactoryMock.Verify(
+            dnsQueryFactory => dnsQueryFactory.GetDnsQueryClient(
+                It.Is<IPAddress>(ipAddress => ipAddress != dnsServerAddress)),
+            Times.Never);
+    }
+
+    [Test]
+    public void SrvQuery_Should_Execute_Srv_Query_For_Specified_Domain()
```

Comment:
I use [strict mock](https://github.com/moq/moq4/wiki/Quickstart#customizing-mock-behavior) which guarantees that no calls other than expected ones can be made. Your example would allow for `queryClient` to be _additionally_ called with other parameters, and I wanted to make sure that could not happen

Reply:
I'm not sure if it's necessary though. I think we should test what we expect. In this case we expect it to be called by our param and we can test agains it. We're also checking if it returns expected results in other tests, so it should be enough.

If there's something we explicitly need to check, for example it's **NOT** called with "any.other.domain", then we can add verify for that as well if needed. By this approach we also miss the part that it needs to be called **only once** for example.

But it's something I just wanted to bring up and discuss. But it's just a suggestion of course. Resolving this comment.

Reply:
> By this approach we also miss the part that it needs to be called only once for example.

"Only once" is a technical detail, it might be more calls in case of retries for example. More important is that there is no call with the wrong parameters, because then you cannot be sure that the output of this method comes from the correct DNS query.

Reply:
I also think that this check isn't very necessary and with the Adil's way the code is a bit simpler but I am good with both.